autokeras/hypermodel/preprocessor.py

```diff
@@ -673,3 +676,233 @@ def set_config(self, config):
         self.gaussian_noise = config['gaussian_noise']
         self.seed = config['seed']
         self.shape = config['shape']
+
+
+def return_zero():
+    return 0
+
+
+class FeatureEngineering(Preprocessor):
+    """A preprocessor block does feature engineering for the data.
+
+    # Arguments
+        column_types: A list of strings. The length of the list should be the same
+            as the number of columns of the data. The strings in the list are
+            specifying the types of the columns. They should either be 'numerical'
+            or 'categorical'.
+        max_columns: Int. The maximum number of columns after feature engineering.
```

Comment:
The meaning of this argument isn't entirely clear. Is this about feature selection? E.g. the max dimensionality of the output space?

Reply:
Yes, it is the maximum number of columns that the output can have. Do you think max_output_dim would be better?

Reply:
There are 2 forms of feature selection one might be doing:

- 1) Removing entire columns that don't seem informative (either numerical or categorical columns)
- 2) Restricting the set of values taken into account for a categorical column

Both of these affect the output dimension. For instance if you have a categorical column with 100 occurrences of value "a", 150 of value "b" and 3 of value "c", you might drop "c" and encode the feature as a 2D binary vector (or even 1D).

"max_columns" makes it sound like you're only going to do 1). Is that the case?

Reply:
We are not removing columns but adding columns. 
So we are not doing 1) but something similar to 2) but not removing the original column but adding the new columns.
At the current stage, we expect max_columns should be larger than the total number of columns, so we would generate more features to reach that limit.